File: Src/hdlc.c
#include "hdlc.h"
#include "stm32f1xx.h"
#include "usart.h"

#define CRC_POLYNOMIAL 0x8408

uint8_t encode[MAX_HDLC_LEN];
uint16_t calculate_crc(uint8_t *data, size_t length) {
    uint16_t crc = 0xFFFF;

    for (size_t i = 0; i < length; i++) {
        crc ^= (uint16_t)data[i];

        for (size_t j = 0; j < 8; j++) {
            if (crc & 0x0001) {
                crc = (crc >> 1) ^ CRC_POLYNOMIAL;
            } else {
                crc = (crc >> 1);
            }
        }
    }

    return crc;
}

void hdlc_encode_and_send(uint8_t *input, size_t length) {
    size_t encode_index = 0;
    hdlc_encode(input, length, encode, &encode_index);
    hdlc_decode_and_save(encode, encode_index);
    Serial_SendArray(encode, encode_index);
}

void hdlc_encode(uint8_t *input, size_t length, uint8_t *output,
                 size_t *output_length) {
    size_t encode_index = 1;
    uint16_t crc = calculate_crc(input, length);
    output[0] = HDLC_FLAG;
    *(input+(length++)) = (uint8_t)(crc & 0xFF);        //LSB
    *(input+(length++)) = (uint8_t)((crc >> 8) & 0xFF); //MSB
    for (size_t i = 0; i < length; i++) {
        if (input[i] == HDLC_FLAG || input[i] == HDLC_ESC) {
            output[encode_index++] = HDLC_ESC;
            output[encode_index++] = input[i] ^ HDLC_ESC_MASK;
        } else {
            output[encode_index++] = input[i];
        }
    }
    output[encode_index++] = HDLC_FLAG;
    *output_length = encode_index;
}

uint8_t decode[MAX_HDLC_LEN];
void hdlc_decode_and_save(uint8_t *input, size_t length) {
    size_t decode_index = 0;
    hdlc_decode(input, length, decode, &decode_index);
}

void hdlc_decode(uint8_t *input, size_t length, uint8_t *output,
                 size_t *output_length) {
    size_t decode_index = 0;
    bool in_frame = false;
    bool crc_match = false;

    for (size_t i = 0; i < length; i++) {
        if (!in_frame) {
            if (input[i] == HDLC_FLAG) {
                in_frame = true;
            }
        } else {
            if (input[i] == HDLC_FLAG) {
                break;
            } else if (input[i] == HDLC_ESC) {
                i++;
                output[decode_index++] = input[i] ^ HDLC_ESC_MASK;
            } else {
                output[decode_index++] = input[i];
            }
        }
    }

    size_t data_fram_len = decode_index - 2;
    uint8_t crc_fram[2];
    crc_fram[0] = output[decode_index - 2];
    crc_fram[1] = output[decode_index - 1];

    uint16_t crc_calculated = calculate_crc(output, data_fram_len);
    uint16_t crc_from_frame = (crc_fram[1] << 8) | crc_fram[0];

    if (crc_calculated == crc_from_frame) {
        crc_match = true;
    }

    if (!crc_match) {
        *output_length = 0;
    } else {
        *output_length = data_fram_len;
    }
}


void test_hdlc() {
    uint8_t data[] = {0x7E, 0x61, 0x7D, 0x5E, 0x7E};
    // hdlc_encode(data, 5);
}

File: Script/hdlc.py
import numpy as np
HDLC_FLAG = 0x7E
HDLC_ESC = 0x7D
HDLC_ESC_MASK = 0x20
CRC_POLYNOMIAL = 0x8408

def calculate_crc(data):
    crc = 0xFFFF
    for byte in data:
        crc ^= byte
        for _ in range(8):
            if crc & 0x0001:
                crc = (crc >> 1) ^ CRC_POLYNOMIAL
            else:
                crc = crc >> 1
    return crc

def hdlc_encode(input):
    encoded = bytearray([HDLC_FLAG])
    crc = calculate_crc(input)
    input_array = bytearray([*input, crc & 0xFF, (crc >> 8) & 0xFF])
    for byte in input_array:
        if byte == HDLC_FLAG or byte == HDLC_ESC:
            encoded.extend([HDLC_ESC, byte ^ HDLC_ESC_MASK])
        else:
            encoded.append(byte)
    encoded.extend([HDLC_FLAG])
    return encoded

def hdlc_decode(input):
    decoded = bytearray()
    in_frame = False
    crc_match = False
    inter = iter(range(len(input)))
    for i in inter:
        byte = input[i]
        if not in_frame:
            if byte == HDLC_FLAG:
                in_frame = True
        else:
            if byte == HDLC_FLAG:
                break
            elif  byte == HDLC_ESC:
                decoded.append(input[i+1] ^ HDLC_ESC_MASK)
                next(inter)
            else:
                decoded.append(byte)
    data_fram = decoded[:-2]
    crc_fram = decoded[-2:]
    crc_calculated = calculate_crc(data_fram)
    crc_from_frame = (crc_fram[1] << 8) |crc_fram[0] 
    if crc_calculated == crc_from_frame:
        crc_match = True
    if not crc_match:
        return None
    else:
        return data_fram

if __name__ == "__main__":
    # Example usage
    data = bytearray([0x7E, 0x61, 0x7D, 0x5E, 0x7E])
    encoded = hdlc_encode(data)
    print(['0x{:02X}'.format(byte) for byte in encoded])
    decoded = hdlc_decode(encoded)
    print(decoded == data)